src/stubs/sm_attest.c

```diff
@@ -0,0 +1,6 @@
+#include "reactive_stubs_support.h"
+
+void SM_ENTRY(SM_NAME) __sm_attest(const uint8_t* challenge, size_t len, uint8_t *result)
+{
+    sancus_tag(challenge, len, result);
```

Comment:
@jovanbulck I tried to fix `sgx_is_outside_enclave`, check out my commit bda59cd

I think the integer overflow issue does not apply in this case, since in `is_buffer_outside_region` the variable `len` is a `size_t`, right?

Reply:
I also am facing another issue here. To sanitize untrusted pointers, I think I have to do something like this (please correct me if I am wrong):
```c
void SM_ENTRY(SM_NAME) __sm_attest(const uint8_t* challenge, size_t len, uint8_t *result)
{
    if( !sancus_is_outside_sm(SM_NAME, challenge, len) ||
        !sancus_is_outside_sm(SM_NAME, result, SANCUS_TAG_SIZE) ) {
      return;
    }

    sancus_tag(challenge, len, result);
}
```

The problem is, I cannot use `SM_NAME` in the code, because I'd get errors when trying to compile the library. In fact, the variable `SM_NAME` does not mean anything at the library's compile time, but is overwritten later on by `linker.py`

So, the thing is, is there a way to circumvent this issue? I am trying to see if there is a function to retrieve the `SancusModule` struct programmatically (or at least, the addresses of the public and secret sections), but I haven't found anything so far. I only found `sancus_get_self_id` to retrieve the ID of the calling SM, but I still have to find out if this can be useful or not.

Reply:
> @jovanbulck I tried to fix `sancus_is_outside_sm`, check out my commit [bda59cd](https://github.com/sancus-tee/sancus-compiler/commit/bda59cd35753ec9857eaa15daef69ad33be05e6f)

Nice, thanks, reviewed

> I think the integer overflow issue does not apply in this case, since in `is_buffer_outside_region` the variable `len` is a `size_t`, right?

No, it does apply(!) A size_t is essentially a uint16_t and it can silently overflow if the attacker passes a large enough buffer towards the end of the address space.

I updated your code to check for overflow as per 5f77836

> Also I wanted to keep `sancus_is_outside_sm` as a macro because otherwise we would either have to (1) pass the SancusModule structure to the inline function by copy (bad), or (2) change it and pass the struct by reference, but risking to break existing code by doing so.

yes agreed, I like your solution!

Reply:
> I also am facing another issue here. To sanitize untrusted pointers, I think I have to do something like this (please correct me if I am wrong):

yes your sanitization code seems correct!

> The problem is, I cannot use `SM_NAME` in the code, because I'd get errors when trying to compile the library. In fact, the variable `SM_NAME` does not mean anything at the library's compile time, but is overwritten later on by `linker.py`
> 
> So, the thing is, is there a way to circumvent this issue? I am trying to see if there is a function to retrieve the `SancusModule` struct programmatically (or at least, the addresses of the public and secret sections), but I haven't found anything so far. I only found `sancus_get_self_id` to retrieve the ID of the calling SM, but I still have to find out if this can be useful or not.

Don't use the SancusModule structs they reside in the _untrusted_ code and can hence not be trusted. Currently the only way to get trusted references to the SM start/end is using the compiler-generated symbols which indeed require the name of the SM.

So this is a generic known issue when compiling SM libraries, there's basically 2 ways tot get around this:

1. parameterize the library sm name but don't compile the library until the client includes it in her SM. This is done by VulCAN eg here: https://github.com/sancus-tee/vulcan/blob/master/can-auth/vulcan.h#L39
2. compile the library with a placeholder SM_NAME=X and later in the linker hack to replace all references to X with the client SM name. This is the approach already done by the reactive toolchain as you describe: https://github.com/sancus-tee/sancus-compiler/blob/master/src/drivers/linker.py#L101

It's all hacky, but I guess it makes sense to stick to (2) here. Ie try to add something like in `get_io_sym_map`:

```python
        '__sm_X__public_start':          '__sm_{}__public_start'.format(sm_name),
       ....
```

hope it helps!

Reply:
another way to avoid having to retrieve the SM name, would be to use a compiler intrinsic. Eg as outline here:

https://github.com/sancus-tee/sancus-compiler/pull/33#discussion_r716683285

This is a bit hacky, but probably neater for the programmer cause you won't have confusion and possible vulnerabilities when you got the name wrong in calls like `sancus_is_outside_sm(SM_NAME, challenge, len)`..

If you want to have a shot at a compiler intrinsic, feel free, but I suggest we simply stick with the current approach. But we should probably think about a proper enclave runtime at some point..

Reply:
I tried to apply method (2), but I had to declare somewhere the variables `__sm_X_{public,secret}_{start,end}` as `extern void*`.

You can check this here: 4830492 . I am not even sure if this is correct or it works, but now the library at least compiles correctly. I will try now running a simple test to check if everything works

Reply:
Yep, I can confirm that the deployment and attestation of an SM works correctly! 

I haven't tested if `sancus_is_outside_sm` actually detects if the input buffer resides in enclave region, this is another test we could write later on maybe

Reply:
okay great! Yes, your code looks good :)

The problem you ran into with the extern declarations, that's exactly why there's the `DECLARE_SM` macro, see here:

https://github.com/sancus-tee/sancus-compiler/blob/master/src/sancus_support/sm_support.h#L24

so you might want to replace that with `DECLARE_SM(X,1234)` I think (the vendor ID doesn't matter at this point)

Reply:
applied in 71d69ac (functionally the same, but a bit neater); hope it doesn't break anything!